### ISSUE TITLE
Centralizing GInteractions into GenomicInteractions

### ISSUE THREAD

**Issue by LTLA** [ComputationalRegulatoryGenomicsICL/GenomicInteractions-old#37]
This continues part of the discussion in [this thread](https://stat.ethz.ch/pipermail/bioc-devel/2019-March/014799.html). The aims relevant to this repository are:

- _GenomicInteractions_ (the package) and `GenomicInteractions` (the class) should be the **only** canonical source for the indexed paired interaction data structure, rather than having the current weird state of affairs with `GInteractions` in _InteractionSet_ and `GenomicInteractions` inheriting from that.
- The `GenomicInteractions` class should inherit from the `IndexedRelations` class in the [_IndexedRelations_](https://github.com/LTLA/IndexedRelations) package. This package generalizes the concept of parallel interactions to any number of partners involving any number of feature sets - pairwise genomic interactions are just a specific case.
- Many of the `GInteractions` methods (e.g., `findOverlaps`) should find homes in _GenomicInteractions_ and operate on `GenomicInteractions`.

@liz-is, @mgperry, is this feasible? The change in the class definition will probably break any code that uses serialized objects, but we can provide `updateObject` methods to get around that. We shouldn't be afraid to rationalize the infrastructure (before it gets too out of hand).

**Comment by mgperry** [comment]
I definitely agree with most of this, this sounds almost like you want to merge the functionality of GenomicInteractions and InteractionSet into one package (with inheritance from existing BioC infrastructure where reasonable) ? I guess in theory most of the work would be on the core class and most of the code should be oblivious to the change.

I don't actually work with 3DGO data any more, so any changes are unlikely to affect me, but I can't speak for either @liz-is or the wider community.

**Comment by LTLA** [comment]
Yes, the idea would be to merge the `GInteractions` features and functionality into _GenomicInteractions_. There will now only be one point of call for all methods and classes relating to "genomic interactions" in the Bioconductor ecosystem (or at least the parts we manage).

Are either of you willing to offer your time to this refactor on the _GenomicsInteraction_ side?  The carrot would be that the underlying class would be able to natively represent instances of interactions between two different "categories", e.g., enhancers and promoters; these would no longer need to be explicitly stored in the same feature set, but can be held separately for greater clarity.

**Comment by LTLA** [comment]
It has begun... see [here](https://github.com/LTLA/GenomicInteractions). The initial plan is to just migrate `GInteractions` code from _InteractionSet_ into _GenomicInteractions_, while also refactoring the `G(enomic)Interactions` class to use `IndexedRelations` as the base. This allows us to cut out a lot of the complexity in the boilerplate methods (e.g., combining, subsetting). 

But any serialized objects will break so some care is required in defining the `updateObject` method - which now needs to work with very old `GenomicInteractions` (before _InteractionSet_) and the current `GenomicInteractions` (based on `GInteractions`) to convert them to the new `GenomicInteractions` (based on `IndexedRelations`).

**Comment by LTLA** [comment]
# Package structure

The existing _InteractionSet_ and _GenomicInteractions_ infrastructure will be split across 4 packages:

- [_IndexedRelations_](https://github.com/LTLA/IndexedRelations), a package to implement the `IndexedRelations` class. This is a general class that supports parallel associations between any `Vector`-like objects. One could use this for storing any types of relationship, e.g., genomic interactions, gene-to-set relations (see kevinrue/unisets#47).
- [_GenomicInteractions_](https://github.com/LTLA/GenomicInteractions) (mark II), which re-implements the `GenomicInteractions` class. This class will be based on the `IndexedRelations` class, allowing us to hand off a lot of the processing logic to _IndexedRelations_ and focus on the genomic aspects of this class. 
- [_fugi_](https://github.com/LTLA/fugi), which implements many of the high-level utilities for manipulating `GenomicInteractions` instances (e.g., plotting, annotating) that previously lived in _GenomicInteractions_ mark I.
- [_InteractionSet_](https://github.com/LTLA/InteractionSet), which implements the _InteractionSet_ class and the _ContactMatrix_ class.

# Code migration plan

- Low-level methods (that work on the data container in a manner that is not specific to genomic intervals) for `GInteractions` will move from _InteractionSet_ to _IndexedRelations_.
- Methods that rely on the genomic aspect of the data container will move from _InteractionSet_ to _GenomicInteractions_, e.g., `findOverlaps`, `boundingBox`.
- Methods related to analysis of genomic interval data (rather than to the structure and manipulation of the `GenomicInteractions` class) will move from _GenomicInteractions_ to _fugi_. This includes all the analysis and plotting code that is currently living in _GenomicInteractions_ mark I.

In essence, we are doing a switcheroo where the `GInteractions` codebase from _InteractionSet_ **becomes** _GenomicInteractions_ mark II, and most of _GenomicInteractions_ mark I becomes _fugi_.

# Responsibilities

I will be responsible for _IndexedRelations_ and _GenomicInteractions_ (mark II). I will create _fugi_ and leave it in a `check`able state, but maintenance responsibility will lie with @mgperry and/or @liz-is until such time that it is stable enough to be handed over to the Bioconductor core team.

And for the all-important citations:

- _GenomicInteractions_ (mark II) will be associated with both the Harmston et al. and Lun et al. papers. Even though it will now only contain code from the latter, the former introduced the class name, so it's probably better to have both papers associated with this package.
- _fugi_ will be associated with the Harmston et al. paper, because that's where the code comes from.

There's probably a paper that could be written for this entire reorganization, if anyone could be bothered.

**Comment by liz-is** [comment]
Hi Aaron,

Sorry for being slow to respond - I'm actually doing a lot of work in the wet lab at the moment, and I won't have time to dedicate to this for at least the next month or two. 

The IndexedRelations package / class sounds like a sensible move, I can see that there's other use cases for that kind of structure so it makes sense to separate it out. 

I'm a bit hesitant about the other reorganisation, though... I'm not sure why the `G(enomic)Interactions` and `InteractionSet` classes need to be in separate packages, other than the fact that packages with those names already exist. 

Also, it seems that making *GenomicInteractions* the home for the class, while moving the analysis and plotting code to a separate package, will be confusing from a user perspective. What's the BioC policy on these kind of rearrangements? Will we have to go through the usual deprecation process for all of the removed functions, which takes three release cycles? That would reduce user confusion but is somewhat frustrating / messy from a maintainer perspective, when the functionality is being moved rather than deprecated. The alternative, I suppose, would be to rename the current *GenomicInteractions* package to *fugi*, if BioC core would agree to that -- although, now I think about it, since renaming tends to be discouraged (if I remember previous bioc-devel discussions correctly), there is probably no process to link a renamed package to previous versions. I guess a new *fugi* package would also have to go through the normal submission process. 

Do you already have an idea of how BioC would handle such a rearrangement, or should we go back to the devel list and ask?

**Comment by LTLA** [comment]
To answer your questions in reverse order:

> Do you already have an idea of how BioC would handle such a rearrangement, or should we go back to the devel list and ask?

I don't know if you remember, but I lived through the renaming of the old `SummarizedExperiment` to `RangedSummarizedExperiment` and the introduction of `SummarizedExperiment0` as a range-free version of the RSE (eventually becoming the new `SummarizedExperiment` class). So it's been done before in exactly the same context, i.e., an intuitive name being occupied by something else.

> Also, it seems that making GenomicInteractions the home for the class, while moving the analysis and plotting code to a separate package, will be confusing from a user perspective. 

If you're referring to the separation of responsibilities into separate packages: I would say that this is a fairly common paradigm across the Bioconductor ecosystem. Low-level methods for interacting with a class are separated from the high-level methods for plotting and analysis. Perhaps it's not so obvious in the 3D-CC field because it's not as hot, but you really notice it in other applications like single-cell data analysis where other developers want the container and methods to manipulate it but they don't want to drag in a whole bunch of other gunk that's not relevant to them. See, for example, the separation of a lot of code from _scater_ into the stand-alone _SingleCellExperiment_.

From the user perspective, I think that the separation of responsibilities will actually clarify things. This _GenomicInteractions_ package is for the container and common methods; this other _fugi_ package is for one type of analysis/plotting; yet another package (e.g., _diffHic_) is for another type of analysis and plotting. The idea is that there is only ever one container, but multiple ways you can use it, without any single analysis being the "canonical" approach - because the appropriate analysis depends too heavily on the scientific context to make a sensible call at the container level.

On a personal level, it makes life easier for both of us as I can just maintain _GenomicInteractions_ and you guys can maintain _fugi_ and the coupling is nice and loose.

If you're referring to the move itself, the long-term change is simple - just `library(fugi)` (or whatever you want to call it). This will give the same behaviour as the current `library(GenomicInteractions)`. In the short-term, I'm hoping to be able to ensure that any code will execute in exactly the same manner, though some of the plotting/analysis functionality will trigger a series of deprecation warnings pushing people to attach _fugi_ rather than _GenomicInterations_. I don't really see a problem with deprecation, this isn't the first large-scale refactoring I've done on someone else's code base (again, _scater_).

> I'm a bit hesitant about the other reorganisation, though... I'm not sure why the G(enomic)Interactions and InteractionSet classes need to be in separate packages, other than the fact that packages with those names already exist.

This touches upon the fundamental problem - _GenomicInteractions_ owns prime real estate in terms of the name, but it doesn't define the guts of the class (which is done by `GInteractions` in _InteractionSet_). This refactoring will make the entire stack of software rational - everything will live where it is supposed to live and developer/user confusion will be minimized. For example, there's now only one base class, avoiding the scenario where some people use `GenomicInteractions` and other people use `GInteractions` (making it impossible to use methods written for `GenomicInteractions` on `GInteractions` subclasses).

**Comment by LTLA** [comment]
The divorce is done. _GenomicInteractions_ now contains the class and some basic/common methods, while _fugi_ contains the high-level methods for analysis, visualization and interpretation. Users and package developers who only used the `GenomicInteractions` class need not do anything; those who used the high-levl functionality in the old _GenomicInteractions_ package simply need to switch from `library(GenomicInteractions)` to `library(fugi)` and most things should just work. 

The major exception is `anchors(..., type="both")` and `anchors(..., type="both")<-`, which has been (i) deprecated in favor of `type=NULL` and (ii) returns/accepts a `Pairs` object rather than a `List`, which is a more natural way of looking at this type of data anyway. A whole host of other behaviors are also deprecated but they should just throw up a host of warnings rather than failing outright.

@mgperry, @liz-is - talk to me. It's all done, the path ahead is clear: I just need your approval to proceed.

**Comment by mgperry** [comment]
Hi Aaron, I hadn't quite appreciated the scope of what you're doing but the package merry-go-round does make a lot of sense. My main concern is that users of the old *GenomicInteractions* will update and then find things missing if they don't keep up with bioc-devel (which, let's be honest, is a pretty exclusive club, so this is ~99% of users). Have you left in functions that, when called, will at least provide an error message instructing the user to `library(fugi)` or would be a headache from a namespace point of view? If not, it might be worth adding an info message to the package start up letting people know (I know this is irritating to some users, but I reckon in this case it's worth it).

Agreed on the `anchors()` change, I've just had re-read the documentation about 5 times to understand what it does so anything is probably an improvement. iirc, `pairs` wasn't on the scene when *GenomicInteractions* was released, in fact it might have been added partly to address exactly these concerns with paired-end sequencing in the like!? I don't remember all of the discussions with Herve.

PS I will try and work my way through some of your code in the next week, I realise now I didn't *really* understand what was going on with `GInteractions` so I've got some catching up to do.

**Comment by LTLA** [comment]
> Have you left in functions that, when called, will at least provide an error message instructing the user to library(fugi) or would be a headache from a namespace point of view? 

Yes, I have thought about this. For various logistical reasons, those functions have been removed but will be put back in, calling the relevant _fugi_ function with a deprecation warning. (Specifically, I've moved all the functions to _fugi_, and the medium-term plan is to have _GenomicInteractions_ `Suggests: fugi` with wrapper functions in the former pointing to the implemented functions in the latter. However, I can't do that right now, because _fugi_ doesn't exist in BioC yet; and submitting _fugi_ would require the new _GenomicInteractions_, which wouldn't be available if the builds fail because of a dependence to _fugi_! Catch-22.)

> in fact it might have been added partly to address exactly these concerns with paired-end sequencing in the like!? 

Think this was from Mike. In hindsight, we could have probably modelled _GenomicInteractions_ off `Pairs`, which would provide an easier implementation at the cost of increased memory usage and some slower algorithms. Our current implementation only demonstrates its benefits when dealing with very large sets of interactions - which, I guess, happens often enough to warrant some optimization.

> I realise now I didn't really understand what was going on

You're not the first person to tell me that :) Just drop me a line if you want some clarification. And/or make an issue or PR if you want to improve the documentation.

**Comment by liz-is** [comment]
I had a chance to look at *GenomicInteractions* today, please see my comments over on https://github.com/LTLA/GenomicInteractions/issues/1.

> Have you left in functions that, when called, will at least provide an error message instructing the user to library(fugi) or would be a headache from a namespace point of view?

This is actually what I meant when I said previously that relocating functionality could be confusing from a user perspective. Your plan for dealing with this sounds fine, and I understand exactly why you can't implement this just yet.

I'll have to look over *fugi* another time, will try to do that asap.
